src/MongoDB.Driver/ClientSessionHandle.cs

```diff
@@ -130,6 +130,7 @@ public IClientSessionHandle Fork()
         /// <inheritdoc />
         public void StartTransaction(TransactionOptions transactionOptions = null)
         {
+            _coreSession.PinnedShardRouter = null;
```

Comment:
Move the setting of PinnedShardRouter to null to inside the StartTransaction method in the CoreSession.cs class.